### DISCUSSION TITLE
Setup with Traefik / Issue with Proxy-Authorization header for services on another host

### DISCUSSION THREAD

**Discussion by voruti** [authelia/authelia#6542] [Troubleshooting] [answered]
I'm unsure on how to properly configure Authelia and Traefik. I (hopefully) set everything up like the docs/guides describe it, but I'm unsure about the effect of some configuration entries:

I configured the Traefik middleware as:
```yml
      traefik.http.middlewares.authelia.forwardauth.address: "http://authelia:9091/api/verify?rd=https%3A%2F%2Fautheliatest.${DOMAIN}%2F"
      traefik.http.middlewares.authelia.forwardauth.trustForwardHeader: "true"
      traefik.http.middlewares.authelia.forwardauth.authResponseHeaders: "Remote-User,Remote-Groups,Remote-Name,Remote-Email"
```
* The `authResponseHeaders` option is only relevant for the service behind the protected URL, am I correct? Since I don't know a service that reads these headers, I can just comment out this option, correct?
* When commenting out the `trustForwardHeader` option, I don't see any change at all. I read the https://www.authelia.com/integration/proxies/fowarded-headers/ docs, but still don't fully understand it. What would be an example where I can see the difference of the presence of this option?

The configuration of `traefik.entryPoints.<name>.forwardedHeaders.trustedIPs` (and `...proxyProtocol...`) can be left empty, when only using a single Traefik instance, a single Authelia instance and running everything on a single host, correct? Else, when using multiple instances and/or hosts, I can input the other hosts into this option, correct?
The right setup for the example below ⤵, would be
1. the IP of host 2 as trustedIPs on host 1

and

2. empty trustedIPs on host 2

correct?

-----

I setup a playground to help myself understand how everything is working: https://github.com/voruti/AutheliaTraefikSetup/tree/f118926f3d89adaf7339a5e1f20e6949f86086e8.
There is probably something wrong with this configuration, because I can't get requests by `Proxy-Authorization` header on different hosts working, like described in https://github.com/voruti/AutheliaTraefikSetup/blob/f118926f3d89adaf7339a5e1f20e6949f86086e8/README.md#results-of-testing. How to properly configure this, or is this a bug?

Is
```yml
      # middleware for Authelia on host1:
      traefik.http.middlewares.authelia.forwardauth.address: "https://autheliatest.${DOMAIN:?error}/api/verify?rd=https%3A%2F%2Fautheliatest.${DOMAIN:?error}%2F"
      traefik.http.middlewares.authelia.forwardauth.trustForwardHeader: "true"
      traefik.http.middlewares.authelia.forwardauth.authResponseHeaders: "Remote-User,Remote-Groups,Remote-Name,Remote-Email"
```
the correct configuration for the Traefik middleware on host 2?

**Comment by voruti** [comment]
Tested the same behaviour on v4.38.0-beta3: https://github.com/voruti/AutheliaTraefikSetup/tree/a4d3229d0fa8edb8fee401e49a169a67dd85c7be ([diff](https://github.com/voruti/AutheliaTraefikSetup/compare/156df3f01bca338f8f9102e77544831e70975ca9...d4f3a24215cccd4983147f0e0ff13ea89ff7ab3e)).

**Comment by james-d-elliott** [comment]
I have had not time to dig through this but will take a look ASAP.

**Comment by james-d-elliott** [comment, marked as answer]
> * The `authResponseHeaders` option is only relevant for the service behind the protected URL, am I correct? Since I don't know a service that reads these headers, I can just comment out this option, correct?

Correct on both counts.



> * When commenting out the `trustForwardHeader` option, I don't see any change at all. I read the https://www.authelia.com/integration/proxies/fowarded-headers/ docs, but still don't fully understand it. What would be an example where I can see the difference of the presence of this option?

This one is actually a bit confusing on Traefiks end unfortunately, basically you can trust forwarded headers in the entrypoint with the trusted IPs option you mentioned and at the middleware option you mentioned, if you don't do both then the forwarded headers in the request are stripped. If you're not using another proxy (like coudflare proxy) then they can generally be removed.The most likely symptom if you need to revisit this is if any is the logged remote IP is wrong.

> There is probably something wrong with this configuration, because I can't get requests by Proxy-Authorization header on different hosts working

I believe traefik strips this header automatically on each hop and they consider a single hop taking place during forward auth, I'll have to double check this particular point. However in the beta if you use the forward auth specific endpoint and configure it to use the `Authorization` header it should work. You'd have to configure the endpoint similar to the below subject to change configuration snippet (making the endpoint `/api/authz/forward-auth` instead of `/api/verify`):

```yaml
server:
  endpoints:
    authz:
      forward-auth:
        implementation: 'ForwardAuth'
        authn_strategies:
          - name: 'HeaderAuthorization'
          - name: 'CookieSession'
```

**Reply by voruti** [reply]
Sooo, using the `Authorization` header with the beta does work :tada:: https://github.com/voruti/AutheliaTraefikSetup/tree/2762eaf2b8f38a5fb6cd6120efc478662e82b4bd.
Is there any disadvantage using that header instead?
Also, I would be interested in understanding why Traefik allows that header but not the other.

**Reply by james-d-elliott** [reply]
Proxy-Authorization is what they call a single hop header, it should be removed after Traefik has handled it. i.e. before it reaches the backend, this is a standard, and it's so headers with credentials don't get passed to every service. At least in my opinion. The reason we used it was this.

However they remove it as soon as they see it, before it is used for forward auth. Which I can understand their reasoning, but they should give an option to users to configure this behavior. I would consider forward auth a relevant destination for such a header especially if traefik doesn't use it for auth.

**Reply by voruti** [reply]
I don't think they remove it (anymore): https://github.com/traefik/traefik/pull/7433.

The problem for me is that Authelia *is* the backend, because I'm calling it from another host: https://github.com/voruti/AutheliaTraefikSetup/blob/4a5d8252f41300b1aaa363905c5792deeea84546/host2/docker-compose.yml#L49.

![](https://www.plantuml.com/plantuml/dpng/VPBFZjD03CRlynIcfzr34ng4X1PKBK92F00toc5sncb6JUp1ur0AsdTdqZHj55RaaZ_-_Fl6d_EOr8hsNMlCAssmGuYj3MIIior2PT8XKBHYEmYaABLr27yCvEl9kiDUk2V_YLiM-2gMKfOYwLXdyIWlrr9ZFHy3xQ6sRSBnMyOZwodYPoRzj-_P6EkK1Vg4OWonHrXzuQImMLsEvBYBJ5aC7tjji0qMR09xULv8LXcETJWCxLcbuzjcKCrcNS6vcW4nkpxFRQkbxOnPpPdLGfLGVWQ7mqJ3B518s3TQRAS2FC3h0ewMD2COh7EOqewcwXCsWJn4uL-d4R0-0zRRgs4cR0eG_D5ZqXrzIE4tFhpR7eCsqA3rAEOgNJPlSdDLW6lG7SuROZbQyKEA_oLj9j8KrdY3TOlr0dzN5MyBA4lYifpp8T2EQc5IuFgwm_T0pF7-HNEu2mHPndAwdnsaZzbmsdisV5ECSO5ojXt-N2HVbbFWD_I5cbH2ZEZDBQRYMnEyH_hUnzkcZ7dCeBvh_m80)

**Reply by voruti** [reply]
I can't seem to find where in the Traefik source code they remove the `Proxy-Authorization` header...

**Reply by voruti** [reply]
I think it's here: https://github.com/golang/go/blob/f19f31f2e7c136a8dae03cbfe4f8ebbb8b54569b/src/net/http/httputil/reverseproxy.go#L294.

**Reply by voruti** [reply]
I think this is one of the "problems": https://github.com/golang/go/issues/50580. Traefik is still using the `Director` function, after which all hop-by-hop headers are removed. It might be possible to switch Traefik to the `Rewrite` function and then implement a custom middleware (for Traefik) that ensures the `Proxy-Authorization` header is kept?

**Reply by dakriy** [reply]
Sorry for the necro bump but some corrections here for anybody who is coming here in the future. The statement:

> However they remove it as soon as they see it, before it is used for forward auth

is not correct. Traefik has not stripped the `Proxy-Authorization` header in forward auth middleware requests since since 2021. It is still stripped from requests going to the service with the attached middleware. [See this PR.](https://github.com/traefik/traefik/pull/7433). [The problem is actually Authelia now](https://www.authelia.com/reference/guides/proxy-authorization/#default-endpoints) since it is no longer a default `Authn` strategy.  Fortunately, you can work around that with the following config. 


```yaml
server:
  endpoints:
    authz:
      forward-auth:
        implementation: 'ForwardAuth'
        authn_strategies:
          - name: 'HeaderProxyAuthorization'
            schemes:
              - 'Basic'
          - name: 'CookieSession'

```

You also need to use the `/api/authz/forward-auth` endpoint in your Traefik forward auth middleware.

@james-d-elliott would it make sense to keep it as a default strategy?

**Reply by james-d-elliott** [reply]
It would not make sense to change the defaults as that'd be a breaking change that also diverges from a normal use case.

Authelia is operating completely normally here as it's configured.

The OP is using the `/api/verify` endpoint which is **_completely unaffected_** by the authz endpoint configs other than to remove the endpoint entirely, it's also worth noting that even today the `/api/verify` endpoint is unchanged and still uses the `Proxy-Authorization` header for basic auth. 

If you look at the [diagram](https://github.com/authelia/authelia/discussions/6542#discussioncomment-8192029) by OP regardless of if the forward auth middleware still strips this header, the second proxy absolutely will.

As such you're completely and undeniably incorrect, and this configuration will not solve the problem. In order to solve the problem the OP should use the new handlers with the `Authorization` header as originally described in the marked answer, change his architecture, or refer to Traefik for a workaround/fix.

**Reply by dakriy** [reply]
Fair enough, GitHub hid the diagram and didn't see it. You are right the second traefik instance will strip the `Proxy-Authorization`. I just wanted to clarify that Traefik does not strip `Proxy-Authorization` for the forward auth calls.

**Comment by voruti** [comment]
With the v4.38.0-beta3 it's possible to choose between the `Proxy-Authorization` and the `Authorization` header. What about a custom option? I.e. any configurable custom string someone wants to use?

**Reply by voruti** [reply]
For now I just put a nginx in front of Authelia to rename the header (using Authelia 4.37.5 stable, not v4.38.0-beta3):
```conf
server {
    listen       80;

    location / {
        proxy_set_header Host $http_host;
        proxy_pass   http://authelia:9091;
        proxy_set_header Proxy-Authorization $http_Authelia_Authorization;
    }
}
```

**Reply by kabegithub** [reply]
If I configured as like below
server 1: traefik --- internet --- server2: traefik --- nginx with this conf ---- authelia 
then I could use authelia authorization head on another server?
